File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/view/RedCircleProgressView.java
package com.module.playways.grab.room.view;


import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.SweepGradient;
import android.util.AttributeSet;
import android.widget.ProgressBar;

import com.common.utils.U;
import com.module.rank.R;


public class RedCircleProgressView extends ProgressBar {

    private Paint mPaint;
    private int mRadius = U.getDisplayUtils().dip2px(23.33f);//半径
    private int strokeWidth = U.getDisplayUtils().dip2px(3.0f);//画笔画线的宽度
    private String str;//时间的字符串

    private RectF rect;

    ValueAnimator mRecordAnimator;

    public RedCircleProgressView(Context context) {
        this(context, null);
    }

    public RedCircleProgressView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RedCircleProgressView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }


    @Override
    protected void onDraw(Canvas canvas) {
        if (mPaint == null) {
            mPaint = new Paint();
        }
        canvas.drawColor(getResources().getColor(R.color.transparent));//画布背景色

        mPaint.setColor(getResources().getColor(R.color.transparent));
        mPaint.setAntiAlias(true);
        mPaint.setStyle(Style.STROKE);//空心
        mPaint.setStrokeWidth(strokeWidth);//线的宽度

        canvas.drawCircle(getWidth() / 2, getHeight() / 2, mRadius, mPaint);
        float[] positions = {0f, 0.5f, 0f};
        int[] colors = {Color.RED, Color.GREEN, Color.BLUE};

        Shader mShader = new SweepGradient(getWidth() / 2, getHeight() / 2,
                new int[]{Color.parseColor("#E9AC1A"),
                        Color.parseColor("#E9AC1A"),
                        Color.parseColor("#E9AC1A"),
                        Color.parseColor("#E9AC1A")
                }, null);

        mPaint.setStrokeJoin(Paint.Join.ROUND);
        mPaint.setShader(mShader);
        mPaint.setDither(true);
        mPaint.setColor(getResources().getColor(R.color.white));
        mPaint.setStrokeWidth(strokeWidth);//覆盖线的宽度
        mPaint.setAntiAlias(true);
        mPaint.setStrokeCap(Paint.Cap.ROUND);
        float sweepAngle = (getMax() - getProgress() * 1.0f) / getMax() * 360;//根据现在值和最大值的百分比计算出弧线现在的度数

        if (rect == null) {
            rect = new RectF();
        }
        rect.left = (getWidth() - mRadius * 2) / 2;
        rect.top = (getHeight() - mRadius * 2) / 2;

        rect.right = rect.left + mRadius * 2;
        rect.bottom = rect.top + mRadius * 2;

        canvas.drawArc(rect, 270, sweepAngle, false, mPaint);

    }

    public void go(long duration){
        cancelAnim();
        setMax(360);
        mRecordAnimator = ValueAnimator.ofInt(0, 360);
        mRecordAnimator.setDuration(duration);
        mRecordAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                int value = (Integer) animation.getAnimatedValue();
                setProgress(value);
            }
        });
        mRecordAnimator.start();
    }

    public void cancelAnim(){
        if(mRecordAnimator != null){
            mRecordAnimator.removeAllUpdateListeners();
            mRecordAnimator.cancel();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        cancelAnim();
    }

    @Override
    protected synchronized void onMeasure(int widthMeasureSpec,
                                          int heightMeasureSpec) {
        //根据半径算出占屏幕的比重，圆环宽，padding相关
        int widthSize = mRadius * 2 + strokeWidth * 2 + getPaddingLeft() + getPaddingRight();
        setMeasuredDimension(widthSize, widthSize);
    }

}

File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/view/RedCircleProgressView.kt
package com.module.playways.grab.room.view


import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.Style
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.SweepGradient
import android.util.AttributeSet
import android.widget.ProgressBar

import com.common.utils.U
import com.common.view.ExPaint
import com.module.playways.R


class RedCircleProgressView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : ProgressBar(context, attrs, defStyle) {

    private val mPaint: Paint = ExPaint()
    private val mRadius = U.getDisplayUtils().dip2px(23.33f)//半径
    private val strokeWidth = U.getDisplayUtils().dip2px(3.0f)//画笔画线的宽度
    private val str: String? = null//时间的字符串

    private var rect: RectF? = null

    private var mRecordAnimator: ValueAnimator? = null

    override fun onDraw(canvas: Canvas) {
        canvas.drawColor(resources.getColor(R.color.transparent))//画布背景色

        mPaint.color = resources.getColor(R.color.transparent)
        mPaint.isAntiAlias = true
        mPaint.style = Style.STROKE//空心
        mPaint.strokeWidth = strokeWidth.toFloat()//线的宽度

        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), mRadius.toFloat(), mPaint!!)
        val positions = floatArrayOf(0f, 0.5f, 0f)
        val colors = intArrayOf(Color.RED, Color.GREEN, Color.BLUE)

        val mShader = SweepGradient((width / 2).toFloat(), (height / 2).toFloat(),
                intArrayOf(Color.parseColor("#E9AC1A"), Color.parseColor("#E9AC1A"), Color.parseColor("#E9AC1A"), Color.parseColor("#E9AC1A")), null)

        mPaint.strokeJoin = Paint.Join.ROUND
        mPaint.shader = mShader
        mPaint.isDither = true
        mPaint.color = resources.getColor(R.color.white)
        mPaint.strokeWidth = strokeWidth.toFloat()//覆盖线的宽度
        mPaint.isAntiAlias = true
        mPaint.strokeCap = Paint.Cap.ROUND
        val sweepAngle = (max - progress * 1.0f) / max * 360//根据现在值和最大值的百分比计算出弧线现在的度数

        if (rect == null) {
            rect = RectF()
        }
        rect!!.left = ((width - mRadius * 2) / 2).toFloat()
        rect!!.top = ((height - mRadius * 2) / 2).toFloat()

        rect!!.right = rect!!.left + mRadius * 2
        rect!!.bottom = rect!!.top + mRadius * 2

        canvas.drawArc(rect!!, 270f, sweepAngle, false, mPaint!!)

    }

    fun go(duration: Long) {
        cancelAnim()
        max = 360
        mRecordAnimator = ValueAnimator.ofInt(0, 360)
        mRecordAnimator!!.duration = duration
        mRecordAnimator!!.addUpdateListener { animation ->
            val value = animation.animatedValue as Int
            progress = value
        }
        mRecordAnimator!!.start()
    }

    private fun cancelAnim() {
        if (mRecordAnimator != null) {
            mRecordAnimator!!.removeAllUpdateListeners()
            mRecordAnimator!!.cancel()
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        cancelAnim()
    }

    @Synchronized
    override fun onMeasure(widthMeasureSpec: Int,
                           heightMeasureSpec: Int) {
        //根据半径算出占屏幕的比重，圆环宽，padding相关
        val widthSize = mRadius * 2 + strokeWidth * 2 + paddingLeft + paddingRight
        setMeasuredDimension(widthSize, widthSize)
    }

}
